File: solutions/zig/sublist/1/sublist.zig
const std = @import("std");

pub const Relation = enum {
    equal,
    sublist,
    superlist,
    unequal,
};

pub fn compare(list_one: []const i32, list_two: []const i32) Relation {
    if (std.mem.eql(i32, list_one, list_two))
        return .equal;
    if (std.mem.indexOf(i32, list_one, list_two) != null)
        return .superlist;
    if (std.mem.indexOf(i32, list_two, list_one) != null)
        return .sublist;
    return .unequal;
}

File: solutions/elixir/sublist/1/lib/sublist.ex
defmodule Sublist do

  defp contains?(_, []), do: true
  defp contains?([], _), do: false
  defp contains?(list, sublist), do: List.starts_with?(list, sublist) or contains?(tl(list), sublist)

  @doc """
  Returns whether the first list is a sublist or a superlist of the second list
  and if not whether it is equal or unequal to the second list.
  """
  def compare(a, a), do: :equal
  def compare(a, b) do
    cond do
      contains?(a, b) -> :superlist
      contains?(b, a) -> :sublist
      true -> :unequal
    end
  end
end
